12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+
+The Merkle tree inner nodes are H(`LnBranch`, lesser-SHA256||greater-SHA256);
+this ordering means that proofs are more compact since left/right is
+inherently determined.
+
+If there are not exactly a power of 2 leaves, then the tree depth will
+be uneven, with the deepest tree on the lowest-order leaves.
+
+e.g. consider the encoding of an `offer` `signature` with TLVs TLV1, TLV2 and TLV3:
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV1) 
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV2) 
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV3) 
+
+Assume L1 < L1nonce, and L2 > L2nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightningoffersignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice: readers will either request an invoice
+(or multiple) or send an invoice based on the offer.  An offer can be much longer-lived than a
+particular invoice, so has some different characteristics; in particular it
+can be recurring, and the amount can be in a non-lightning currency.  It's
+also designed for compactness, to easily fit inside a QR code.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 6 (`currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 20 (`issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 22 (`quantity_min`)
+    2. data:
+        * [`tu64`:`min`]
+    1. type: 24 (`quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 30 (`node_id`)
+    2. data:
+        * [`point32`:`node_id`]
```

Comment:
Why do we use a 32 bytes public key here and not a 33 bytes one like everywhere else? I understand that Schnorr's signatures don't need the extra byte but the rest of our system needs it.

Reply:
x-only pubkeys; I figured when I wrote this that the network would end up going there (gossipv2 works this way, for example)/

Meanwhile in my implementation I look for both 0x2 and 0x3 variants when trying to route messages.

Reply:
Is this assumption still the case? If not, using 33 bytes seems more desirable as it would allow for reusing existing code paths.

Reply:
There will be a new commit switching to 33 byte pubkeys: turns out x-only maximalism was wrong :(